### ISSUE TITLE
Incomplete todo file generation?

### ISSUE THREAD

**Issue by rogerkk** [troessner/reek#914]
I'm upgrading from reek 1.x, and a lot of stuff has changes since I last touched this gem, and I got a lot of new warnings.

I was glad to discover the new todo file generation, but it seems like the file generated is somewhat incomplete. I'm not sure if I've missed something here or if this is actually a bug, but the user experience here is a bit confusing so I'm hoping this issue is useful.

Here's what I'm seeing:
```
$ reek -t app/

'.todo.reek' generated! You can now use this as a starting point for your configuration.
$ wc -l .todo.reek
1094 .todo.reek
$ reek app/
...
47 total warnings
```

The warnings are of different types, like Attribute, TooManyInstanceVariables, TooManyStatements, IrresponsibleModule etc.

Is this a known issue? If so, please let me know what info would be most relevant to help you debug it.

**Comment by troessner** [comment]
Can't really reproduce this:

Sample file:

```
  ( $ ) cat foo.rb
class Foo
  attr_accessor :bar

  def hola(arg = true)
    bla arg
    bla arg
  end
end
```

Running Reek:

```
( $ ) reek foo.rb
foo.rb -- 4 warnings:
  [2]:Attribute: Foo#bar is a writable attribute [https://github.com/troessner/reek/blob/master/docs/Attribute.md]
  [4]:BooleanParameter: Foo#hola has boolean parameter 'arg' [https://github.com/troessner/reek/blob/master/docs/Boolean-Parameter.md]
  [5, 6]:DuplicateMethodCall: Foo#hola calls bla arg 2 times [https://github.com/troessner/reek/blob/master/docs/Duplicate-Method-Call.md]
  [1]:IrresponsibleModule: Foo has no descriptive comment [https://github.com/troessner/reek/blob/master/docs/Irresponsible-Module.md]
```

Generating the todo file:

```
  ( $ ) reek --todo foo.rb

'.todo.reek' generated! You can now use this as a starting point for your configuration.
```

And running Reek again:


```
  ( $ ) reek -c .todo.reek foo.rb
  ( $ )
```

Can you please give me a reproducible example similar to my example?

**Comment by mvz** [comment]
@rogerkk is another `.reek` file present in the directory?

**Comment by rogerkk** [comment]
> Can you please give me a reproducible example similar to my example?

I'm sorry. I've tried but I haven't been able to reproduce this with a constructed example.

> is another .reek file present in the directory?

I've deleted the original `.reek ` file. I've also ran it with the -c paramater like @troessner mentions: `reek -c .todo.reek app/`

**Comment by rogerkk** [comment]
I've noticed another strange behavior, though. It seems that when generating the todo file, reek actually respects an existing reek file. Is this intended?

```
$ reek -t app/; wc -l .todo.reek 

'.todo.reek' generated! You can now use this as a starting point for your configuration.
1144 .todo.reek
$ reek -t app/; wc -l .todo.reek 

'.todo.reek' generated! You can now use this as a starting point for your configuration.
9 .todo.reek
```

This is two consecutive runs on the same app, where it seems that a handful of rules not caught in the first run makes out the entire todo file after the second run.

**Comment by troessner** [comment]
>> It seems that when generating the todo file, reek actually respects an existing reek file. Is this intended?

Yes and no 😆 

Yes, because this would be a cool feature. 
No, because Reek actually only allows one configuration file. So if you have already one and then generate a .todo.yml it's actually impossible to use it the right way (without moving the whole content of the .todo.yml to the original configuration file).

I believe we have a couple of conceptual kinks to iron out here @mvz and @chastell and then revise this feature a little. 

>> $ reek -t app/; wc -l .todo.reek 
'.todo.reek' generated! You can now use this as a starting point for your configuration.
9 .todo.reek

There it is, our bug! Can you please paste the file here and some (not all) of the corresponding code snippets?

**Comment by mvz** [comment]
> I believe we have a couple of conceptual kinks to iron out here

I agree. I'll ponder this for a bit.

**Comment by rogerkk** [comment]
> There it is, our bug! Can you please paste the file here and some (not all) of the corresponding code snippets?

Okay, but I have to warn you. It's not pretty. :blush:

```
$ cat .todo.reek 
---
TooManyStatements:
  exclude:
  - DentistSearch#initialize
  - RegularSearch#initialize
  - SpecialistSearch#initialize
  - WriteReviewSearch#initialize
  - Reactions::DataRowParser#initialize
  - Reactions::LineParser#initialize
```

The *Search#initializer methods are pretty similar, so perhaps one is enough.

DentistSearch#initialize:
``` ruby
  def initialize(query)
    super(query)

    @service_type_filter = @query.params[:service_type]
    @appointment_booking_filter = @query.params[:booking]
    @procedure_id = @query.params[:procedure]

    sphinx_filter :service_type_filter,
                  Search::ServiceTypeFilter.new(@service_type_filter).value
    sphinx_filter :appointment_booking_filter, @appointment_booking_filter ? true : nil

    sphinx_filter :display_in_main_search, true
    sphinx_option :classes, [Dentist, DentistPractice]

    sphinx_option :sql, include: [{ person_organizations:
                                    { organization:
                                      [{ zip_code:
                                        { zip_place:
                                          { municipality: :county } } },
                                       :borough]
                                    }
                                  }]

    # HACK: Override :order option originally set in RegularSearch, because
    # #sort_field would not know of the @procedure_id variable when
    # RegularSearch#initialize gets executed and thus it would never
    # actually order by price for any other procedure than the default.
    sphinx_option :order, SortOrder.new(sort_field, @query.params[:direction]).to_sql
  end
```

Reactions::LineParser#initialize:
```ruby
    def initialize(data_row, logger)
      @data_row = data_row

      # to make validators work
      @case_number = @data_row[:case_number]
      @hpr_number = @data_row[:hpr_number]
      @appealed_text = @data_row[:appealed_text]
      @reaction_type = @data_row[:reaction_type]

      @logger = logger
    end
```

Let me know if you need more!

**Comment by troessner** [comment]
I'm afraid not:

````
  ( $ ) cat line_parser.rb
class SpecialistSearch
  def initialize(data_row, logger)
    @data_row = data_row

    # to make validators work
    @case_number = @data_row[:case_number]
    @hpr_number = @data_row[:hpr_number]
    @appealed_text = @data_row[:appealed_text]
    @reaction_type = @data_row[:reaction_type]

    @logger = logger
  end
end

  ( $ ) reek line_parser.rb
line_parser.rb -- 2 warnings:
  [1]:IrresponsibleModule: SpecialistSearch has no descriptive comment [https://github.com/troessner/reek/blob/master/docs/Irresponsible-Module.md]
  [1]:TooManyInstanceVariables: SpecialistSearch has at least 6 instance variables [https://github.com/troessner/reek/blob/master/docs/Too-Many-Instance-Variables.md]

  ( $ ) reek --todo line_parser.rb

'.todo.reek' generated! You can now use this as a starting point for your configuration.

  ( $ ) reek -c .todo.reek line_parser.rb
  ( $ )
```

**Comment by rogerkk** [comment]
I've tried copying these 9 files to a separate folder and running `reek -t` on them. This generates a todo file that eliminates all warnings. :(

Thus I guess we have no simple way of reproducing this without us exposing large chunks of our codebase. Or perhaps you have some other idea?

**Comment by troessner** [comment]
>> Thus I guess we have no simple way of reproducing this without us exposing large chunks of our codebase.

Yep, that's probably a no-go.

>> Or perhaps you have some other idea?

Hmmmm. Do you do *weird* things with the problematic classes or modules? E.g. dynamically re-defining them or something? Is there anything in LineParser that seems off to you compared with regular PORO's?

**Comment by mvz** [comment]
@troessner I don't think @rogerkk's specific code is relevant. It seems to me the following happens:
* There is a file `foo.reek`, containing at least exclusions for `TooManyStatements`.
* `reek --todo` runs, uses `foo.reek` as a base configuration, and doesn't generate `TooManyStatements` exclusions.
* `reek --todo` runs again, but uses `.todo.reek` as a base configuration, and now *only* generates `TooManyStatements` exclusions.

**Comment by troessner** [comment]
```
There is a file foo.reek, containing at least exclusions for TooManyStatements.
reek --todo runs, uses foo.reek as a base configuration, and doesn't generate TooManyStatements exclusions.
reek --todo runs again, but uses .todo.reek as a base configuration, and now only generates TooManyStatements exclusions.
```

But contrasting this with 

```
$ reek -t app/; wc -l .todo.reek 

'.todo.reek' generated! You can now use this as a starting point for your configuration.
1144 .todo.reek
$ reek -t app/; wc -l .todo.reek 

'.todo.reek' generated! You can now use this as a starting point for your configuration.
9 .todo.reek
```

would mean that the existing Reek config would have contained **only** those very 9 exceptions and ignoring more than 1000 others. Seems a bit absurd but you'll never know 😆 

@rogerkk can you please cure of from our confusion and post the original Reek config you were using?

**Comment by rogerkk** [comment]
@troessner I actually deleted my existing reek config before running the above commands. A `find . -name *.reek` doesn't return anything, so the first run is thus a "clean sheet". :/

**Comment by troessner** [comment]
>> I actually deleted my existing reek config before running the above commands. A find . -name *.reek doesn't return anything, so the first run should be a "clean sheet".

That would actually contradict your theory @mvz, wouldn't it?

**Comment by mvz** [comment]
> That would actually contradict your theory @mvz, wouldn't it?

Yes, it does :disappointed:.

Except `find -name *.reek` returns `zsh: no matches found: *.reek` in my shell, so find doesn't even run. @rogerkk can you/did you try `find -name '*.reek'` please (that is, with the quotes)?

**Comment by jhubert** [comment]
Fwiw, I've experienced this as well. No prior reek file.

```
⌘ 15:20:34 zipline-light (763-teams-issue *)$ gem install reek
Fetching: parser-2.3.1.0.gem (100%)
Successfully installed parser-2.3.1.0
Fetching: equalizer-0.0.11.gem (100%)
Successfully installed equalizer-0.0.11
Fetching: descendants_tracker-0.0.4.gem (100%)
Successfully installed descendants_tracker-0.0.4
Fetching: coercible-1.0.0.gem (100%)
Successfully installed coercible-1.0.0
Fetching: ice_nine-0.11.2.gem (100%)
Successfully installed ice_nine-0.11.2
Fetching: axiom-types-0.1.1.gem (100%)
Successfully installed axiom-types-0.1.1
Fetching: virtus-1.0.5.gem (100%)
Successfully installed virtus-1.0.5
Fetching: codeclimate-engine-rb-0.3.1.gem (100%)
Successfully installed codeclimate-engine-rb-0.3.1
Fetching: reek-4.0.2.gem (100%)
Successfully installed reek-4.0.2
9 gems installed
⌘ 20:00:08 zipline-light (master u=)$ reek --todo

'.todo.reek' generated! You can now use this as a starting point for your configuration.
⌘ 20:00:18 zipline-light (master u=)$ reek
app/models/communication_target_mapper.rb -- 1 warning:
  [42]:TooManyStatements: CommunicationTargetMapper#initialize has approx 7 statements [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
app/models/summary.rb -- 1 warning:
  [17]:TooManyStatements: Summary#initialize has approx 8 statements [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
app/models/tracking/team/query.rb -- 1 warning:
  [7]:TooManyStatements: Tracking::Team::Query#initialize has approx 7 statements [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
app/models/upload_policy.rb -- 1 warning:
  [4]:TooManyStatements: UploadPolicy#initialize has approx 6 statements [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
app/models/zipline_digest.rb -- 1 warning:
  [5]:TooManyStatements: ZiplineDigest#initialize has approx 8 statements [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
5 total warnings
```

**Comment by jhubert** [comment]
In the above, it looks like all of the items are on the `initialize` statements, which seems doubly weird to me because that seems counter to [the docs](https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md):

    By default, initialize is not checked for length; any class's constructor can be as long as necessary.

It may be that the todo file generates configurations that change the default behavior, so when it's run again the changed behavior finds new smells.

**Comment by jhubert** [comment]
Aaaannd, yes. That's what it is.

The `TooManyStatements` config gets set to:

```yaml
TooManyStatements:
  exclude:
  - MyClassName#method
```
which overwrites the default exclude config for `TooManyStatements`:

```yaml
TooManyStatements:
  exclude:
  - initialize
```

**Comment by mvz** [comment]
Good catch @jhubert, thanks!

**Comment by troessner** [comment]
Excellent catch! I can look into this, but it will take me a while since i am very busy at the moment. 

@mvz are you up for it or are you alsk blocked?

**Comment by jhubert** [comment]
What do you think the correct behavior should be?

1. the generated todo.reek file should include the default excludes
2. the reek default option of ignoring initialize methods for TooManyStatements becomes a behavior, so that initialize is always included in the `except` list
3. the defaults should be deep merged with the configuration only in the case that it's a `.todo.reek` file
4. something else?

My first instinct was #1.

**Comment by troessner** [comment]
I think it should be (1) as well.

**Comment by mvz** [comment]
I'm not at my computer for the coming week, so let's discuss again who can pick this up in a week.

**Comment by mvz** [comment]
I can probably take a stab at this over the weekend.

**Comment by troessner** [comment]
🎂

**Comment by mvz** [comment]
I have another option: The config values should always be merged with the defaults. This solves another problem with the config file, which is that you have to add those default exclusions yourself as well if you want to exclude something else.

This would mean we need another mechanism to 'undo' an exclude. We could introduce an 'include' for that, which would be applied first.

**Comment by mvz** [comment]
I'll go with option 1 first, since the alternative would be a breaking change.

**Comment by troessner** [comment]
>> This would mean we need another mechanism to 'undo' an exclude. We could introduce an 'include' for that, which would be applied first.

That sounds complicated. I can see additional bug reports coming in 😆 
What do you think about a SmellDetector flag like

`ignore_default_configuration: true | false`

with `false` being the default? Wouldn't that take out some of the complexity and yet achieve the same?
Also this seems like something that is easier to explain to users.

**Comment by chastell** [comment]
FWIW: RuboCop has an `inherit_from` clause, and my typical usage is to have a `.rubocop.yml` starting with `inherit_from: .rubocop_todo.yml` (followed by disabling the cops I disagree with and tweaking other cops’ defaults); this way `.rubocop_todo.yml` ends up having only things that are not explicitly whitelisted. Maybe a similar approach would work for Reek?

**Comment by troessner** [comment]
I know rubocops `inherit_from` and i think it's awesome. For rubocop, not necessarily for Reek. Both gems are really different and I believe Reek can and should have a way simpler configuration mechanism.
wdyt about my suggestion from above @chastell ?

**Comment by chastell** [comment]
I don’t see as much difference between configuring RuboCop and Reek, but I agree we should go with something simple first and if that’s enough then that’s even better. :smile: For me `ignore_default_configuration: true` is just another way of writing `inherit_from: defaults`, so I’m all for starting with that.

**Comment by troessner** [comment]
Fixed and released with 4.0.3 ;)
